x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/ChangePoint.java

```diff
@@ -0,0 +1,207 @@
+    }
+
+    @Override
+    public List<Attribute> output() {
+        if (output == null) {
+            output = NamedExpressions.mergeOutputAttributes(List.of(targetType, targetPvalue), child().output());
+        }
+        return output;
+    }
+
+    public NamedExpression value() {
+        return value;
+    }
+
+    public NamedExpression key() {
+        return key;
+    }
+
+    public Attribute targetType() {
+        return targetType;
+    }
+
+    public Attribute targetPvalue() {
+        return targetPvalue;
+    }
+
+    @Override
+    public List<Attribute> generatedAttributes() {
+        return List.of(targetType, targetPvalue);
+    }
+
+    @Override
+    public ChangePoint withGeneratedNames(List<String> newNames) {
+        checkNumberOfNewNames(newNames);
+        Attribute newTargetType;
+        if (targetType.name().equals(newNames.get(0))) {
+            newTargetType = targetType;
+        } else {
+            newTargetType = targetType.withName(newNames.get(0)).withId(new NameId());
+        }
+        Attribute newTargetPvalue;
+        if (targetPvalue.name().equals(newNames.get(1))) {
+            newTargetPvalue = targetPvalue;
+        } else {
+            newTargetPvalue = targetPvalue.withName(newNames.get(1)).withId(new NameId());
+        }
+        return new ChangePoint(source(), child(), value, key, newTargetType, newTargetPvalue);
+    }
+
+    @Override
+    protected AttributeSet computeReferences() {
+        return Expressions.references(List.of(key, value));
+    }
+
+    @Override
+    public boolean expressionsResolved() {
+        return value.resolved() && key.resolved();
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(value, key, targetType, targetPvalue, child());
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+        if (obj == null || getClass() != obj.getClass()) {
+            return false;
+        }
+        ChangePoint other = (ChangePoint) obj;
+        return Objects.equals(child(), other.child())
+            && Objects.equals(value, other.value)
+            && Objects.equals(key, other.key)
+            && Objects.equals(targetType, other.targetType)
+            && Objects.equals(targetPvalue, other.targetPvalue);
+    }
+
+    private Order order() {
+        return new Order(source(), key, Order.OrderDirection.ASC, Order.NullsPosition.ANY);
+    }
+
+    @Override
+    public LogicalPlan surrogate() {
+        // ChangePoint should always run on the coordinating node after the data is collected
+        // in sorted order. This is enforced by adding OrderBy here.
+        // Furthermore, ChangePoint should be called with at most 1000 data points. That's
+        // enforced by the Limits here. The first Limit of N+1 data points is necessary to
+        // generate a possible warning, the second Limit of N is to truncate the output.
```

Comment:
nit: I'd personally put this in ChangePoint's javadoc as that's quite important to know.